runtime/compiler/optimizer/J9TransformUtil.cpp

```diff
@@ -969,6 +997,267 @@ J9::TransformUtil::canFoldStaticFinalField(TR::Compilation *comp, TR::Node* node
+   }
+
+static bool isVarHandleFolding(TR::Compilation* comp, TR_OpaqueClassBlock* declaringClass, char* fieldSignature, int32_t fieldSigLength)
+   {
+   if (comp->getMethodSymbol()->hasMethodHandleInvokes()
+       && !TR::Compiler->cls.classHasIllegalStaticFinalFieldModification(declaringClass))
+      {
+      if (fieldSigLength == 28 && !strncmp(fieldSignature, "Ljava/lang/invoke/VarHandle;", 28))
+         return true;
+      }
+
+   return false;
+   }
+
+/** \brief
+ *     Try to fold var handle static final field with protection
+ *
+ *  \param opt
+ *     The current optimization object.
+ *
+ *  \param currentTree
+ *     The tree with the load of static final field.
+ *
+ *  \param node
+ *     The node which is a load of a static final field.
+ */
+bool J9::TransformUtil::attemptVarHandleStaticFinalFieldFolding(TR::Optimization* opt, TR::TreeTop * currentTree, TR::Node *node)
+   {
+   return J9::TransformUtil::attemptStaticFinalFieldFoldingImpl(opt, currentTree, node, true);
+   }
+
+/** \brief
+ *     Try to fold generic static final field with protection
+ *
+ *  \param opt
+ *     The current optimization object.
+ *
+ *  \param currentTree
+ *     The tree with the load of static final field.
+ *
+ *  \param node
+ *     The node which is a load of a static final field.
+ */
+bool J9::TransformUtil::attemptGenericStaticFinalFieldFolding(TR::Optimization* opt, TR::TreeTop * currentTree, TR::Node *node)
+   {
+   return J9::TransformUtil::attemptStaticFinalFieldFoldingImpl(opt, currentTree, node, false);
+   }
+
+/** \brief
+ *     Try to fold static final field with protection
+ *
+ *  \param opt
+ *     The current optimization object.
+ *
+ *  \param currentTree
+ *     The tree with the load of static final field.
+ *
+ *  \param node
+ *     The node which is a load of a static final field.
+ *
+ *  \param varHandleOnly
+ *     True if only folding varHandle static final fields.
+ *     Faslse if folding all static final fileds.
+ */
+bool J9::TransformUtil::attemptStaticFinalFieldFoldingImpl(TR::Optimization* opt, TR::TreeTop * currentTree, TR::Node *node, bool varHandleOnly)
+   {
+   TR::Compilation* comp = opt->comp();
+   // first attempt folding reliable fields
+   if (J9::TransformUtil::foldReliableStaticFinalField(comp, node))
+      {
+      if (opt->trace())
+         traceMsg(comp, "SFFF fold reliable at node %p\n", node);
+      return true;
+      }
+
+   // try folding regular static final fields
+   if (comp->getOption(TR_DisableGuardedStaticFinalFieldFolding))
+      {
+      return false;
+      }
+
+   if (!comp->supportsInduceOSR()
+       || !comp->isOSRTransitionTarget(TR::postExecutionOSR)
+       || comp->getOSRMode() != TR::voluntaryOSR)
+      {
+      return false;
+      }
+
+   if (skipFinalFieldFoldingInBlock(comp, currentTree->getEnclosingBlock())
+       || safeToAddFearPointAt(opt, currentTree) != TR_yes
+       || J9::TransformUtil::canFoldStaticFinalField(comp, node) != TR_maybe)
```

Comment:
I didn't think about this when I wrote the code, but can we move `canFoldStaticFinalField` before `safeToAddFearPointAt` so that we do the cheaper checks first?

Reply:
Addressed.